### PR TITLE
[FIX] Fetch json key from item serializer if empty collection is pass…

### PR DESCRIPTION
**Pull request by RomanKapitonov**
…ed to collection serializer and each_searializer is specified.

### CODE REVIEW

**Comment by RomanKapitonov** [comment]
Fixes https://github.com/rails-api/active_model_serializers/issues/1536

File: test/fixtures/poro.rb [hunk 1]
```diff
@@ -164,6 +164,12 @@ def json_key
   end
 end
 
+MessagesSerializer = Class.new(ActiveModel::Serializer) do
+  def json_key
+    'messages'
+  end
+end
```
Comment (bf4) [inline test/fixtures/poro.rb]: would you mind moving this to the test file? poro's is a little cluttered

File: lib/active_model/serializer/collection_serializer.rb [hunk 2]
```diff
@@ -39,8 +45,15 @@ def paginated?
       private
 
       def derived_root
-        key = serializers.first.try(:json_key) || object.try(:name).try(:underscore)
-        key.try(:pluralize)
+        serializers.first.try(:json_key).try(:pluralize)
```
Comment (bf4) [inline lib/active_model/serializer/collection_serializer.rb]: we really need to replace `Serializer#json_key` with `Serializer._type`...

File: lib/active_model/serializer/collection_serializer.rb [hunk 3]
```diff
@@ -10,8 +10,14 @@ class CollectionSerializer
       def initialize(resources, options = {})
         @root = options[:root]
```
Comment (bf4) [inline lib/active_model/serializer/collection_serializer.rb]: I'm thinking this should really be `@root = collection_root(options[:root], options[:serializer])`

and something like (assuming root and json_key should behave the same... I don't remember)

```ruby
private

def collection_root(explicit_root, explicit_each_serializer)
  explicit_root || derived_root(explicit_each_serializer)
end

def derived_root(explicit_each_serializer)
 serializers.first.try!(:json_key).try!(:pluralize) ||
    (resources.empty && explicit_each_serializer.try!(:allocate).try!(:json_key).try!(:pluralize)) ||
    object.try!(:name).try!(:underscore).try!(:pluralize)
end
```

aside: all these `try` make my head hurt

Comment (groyoh) [inline lib/active_model/serializer/collection_serializer.rb]: :+1: I also don't really enjoy all these `try`.

File: lib/active_model/serializer/collection_serializer.rb [hunk 4]
```diff
@@ -39,8 +45,15 @@ def paginated?
       private
 
       def derived_root
-        key = serializers.first.try(:json_key) || object.try(:name).try(:underscore)
-        key.try(:pluralize)
+        serializers.first.try(:json_key).try(:pluralize)
+      end
+
+      def default_root
+        object.try(:name).try(:underscore).try(:pluralize)
```
Comment (groyoh) [inline lib/active_model/serializer/collection_serializer.rb]: What about:
```ruby
def default_root
  return unless object.respond_to?(:name)
  object.name.to_s.underscore.pluralize
end

File: lib/active_model/serializer/collection_serializer.rb (hunk 2 above)
Comment (groyoh) [inline lib/active_model/serializer/collection_serializer.rb]: Here what about:

```ruby
def derived_root
  return unless serializers.empty?
  serializers.first.json_key.to_s.pluralize
end
```

Makes it more readable IMO.

File: lib/active_model/serializer/collection_serializer.rb [hunk 5]
```diff
@@ -10,8 +10,14 @@ class CollectionSerializer
       def initialize(resources, options = {})
         @root = options[:root]
         @object = resources
+
+        serializer_context_class = options.fetch(:serializer_context_class, ActiveModel::Serializer)
+
+        if resources.blank? && options[:serializer]
+          @each_serializer = options[:serializer]
```
Comment (groyoh) [inline lib/active_model/serializer/collection_serializer.rb]: `@each_serializer` is too much linked to the controller option name IMO. I'd rather call it `@resource_serializer` or simply `@serializer`.

Comment (bf4) [inline lib/active_model/serializer/collection_serializer.rb]: Or explicit _collection_serializer

File: lib/active_model/serializer/collection_serializer.rb (hunk 3 above)
Comment (bf4) [inline lib/active_model/serializer/collection_serializer.rb]: Right now, the only scenario tested is essentially:

```diff
@root = collection_root(options)

# condenses `root || derived_root`
def collection_root(options)
  if root = options[:root]
    root
 else
-   key = serializers.first.try(:json_key) || object.try(:name).try(:underscore)
+   key = serializers.first.try(:json_key) || options[:serializer].try!(:allocate).try!(:json_key).try!(pluralize) || object.try(:name).try(:underscore)
    key.try(:pluralize)
  end
end
```

(or a more thorough impl would be something like)

```ruby
@root = collection_root(options)

# condenses `root || derived_root`
def collection_root(options)
  if root = options[:root]
    root
 else
   key = object.empty? && (serializer_class = options[:serializer]) && (serializer_class.allocate.json_key.try!(:pluralize) || serializer_class._type) # shouldn't need the allocate here
    key ||= (serializer = serializers.first) && serializer.json_key
    key ||= object.any? && object.name.underscore
    key ||= '' # oh no! we should never be here. empty collection and no serializer option
    key.pluralize
  end
end
```
I'm also realizing we're inconsistent with object vs. resources.  We should probably always refer to it as object (which is the serializer api for the thing it encapsulates).

**Comment by bf4** [comment]
@RomanKapitonov are you planning on coming back to this?

File: test/fixtures/poro.rb (hunk 1 above)
Comment (NullVoxPopuli) [inline test/fixtures/poro.rb]: :+1:

**Comment by NullVoxPopuli** [comment]
@RomanKapitonov are you planning on helping resolve the PR Review comments? :-\

File: lib/active_model/serializer/collection_serializer.rb (hunk 3 above)
Comment (groyoh) [inline lib/active_model/serializer/collection_serializer.rb]: @bf4 @NullVoxPopuli  @remear  I really think that `allocate` is not the right solution here. `json_key` by default use `root` or the `object` (https://github.com/rails-api/active_model_serializers/blob/master/lib/active_model/serializer.rb#L185). Neither of these would be available and might result in a bug if the `json_key` method is not overidden. That might also break for polymorphic collections.

I stay on my position that the prefered solution here would be the ones described at https://github.com/rails-api/active_model_serializers/issues/1536#issuecomment-188434945. Or even subclass the collection serializer and add its own root handling.

**Comment by bf4** [comment]
Followed up with #1618

**Pull request closed without merging**